controlplane/pkg/common/excluded_prefixes_service.go

```diff
@@ -0,0 +1,111 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/pkg/errors"
+
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/connection"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/connectioncontext"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/networkservice"
+	"github.com/networkservicemesh/networkservicemesh/sdk/prefix_pool"
+)
+
+// ConnectionService makes basic Mechanism selection for the incoming connection
+type excludedPrefixesService struct {
+	prefixes prefix_pool.PrefixPool
+}
+
+// NewExcludedPrefixesService -  creates a service to select endpoint.
+func NewExcludedPrefixesService() networkservice.NetworkServiceServer {
+	return NewExcludedPrefixesServiceFromPath(prefix_pool.PrefixesFilePathDefault)
+}
+
+// NewExcludedPrefixesService -  creates a service to select endpoint.
+func NewExcludedPrefixesServiceFromPath(configPath string) networkservice.NetworkServiceServer {
+	return &excludedPrefixesService{
+		prefixes: prefix_pool.NewPrefixPoolReader(configPath),
+	}
+}
+
+func (eps *excludedPrefixesService) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*connection.Connection, error) {
+	logger := Log(ctx)
+
+	if request.GetConnection() == nil {
+		return nil, errors.Errorf("request's connection cannot be empty")
+	}
+
+	// 7.1.6 Update Request with exclude_prefixes, etc
+	requestNext := request.Clone()
+	conn := requestNext.Connection
+	if conn.Context == nil {
+		conn.Context = &connectioncontext.ConnectionContext{}
+	}
+	if conn.Context.IpContext == nil {
+		conn.Context.IpContext = &connectioncontext.IPContext{}
+	}
+	prefixes := eps.prefixes.GetPrefixes()
+	logger.Infof("ExcludedPrefixesService: adding excluded prefixes to connection: %v", prefixes)
+	ipCtx := conn.Context.IpContext
+	ipCtx.ExcludedPrefixes = append(ipCtx.GetExcludedPrefixes(), prefixes...)
+
+	conn, err := ProcessNext(ctx, requestNext)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = eps.validateConnection(conn); err != nil {
+		logger.Errorf("ExcludedPrefixesService: connection is invalid: %v", err)
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+func (eps *excludedPrefixesService) validateConnection(conn *connection.Connection) error {
+	if err := conn.IsComplete(); err != nil {
+		return err
+	}
+
+	ipCtx := conn.GetContext().GetIpContext()
+	if err := eps.validateIPAddress(ipCtx.GetSrcIpAddr(), "srcIP"); err != nil {
+		return err
+	}
+
+	return eps.validateIPAddress(ipCtx.GetDstIpAddr(), "dstIP")
+}
+
+func (eps *excludedPrefixesService) validateIPAddress(ip, ipName string) error {
+	if ip == "" {
+		return nil
+	}
+	intersect, err := eps.prefixes.Intersect(ip)
+	if err != nil {
+		return err
+	}
+	if intersect {
+		return errors.Errorf("%s intersects excluded prefixes list", ipName)
```

Comment:
Error is not provide information about actual IPs to understand what is required to fix it for customer. Please enhance error messaging with current list of excluded IPs and one provided.

Reply:
- [ ] enhance error messaging with current list of excluded IPs and one provided